### ISSUE TITLE
[Question] multiple value ranges

### ISSUE THREAD

**Issue by aledeg** [nelmio/alice#486]
I am not sure if the title make any sense. Please rephrase it if needed.
With explanation, it would be easier I think.

I want to create many object for testing within a set of value. I want to create all different contacts were the lastname comes from a list, the firstname and the dob as well.
I would like to write something like:
```yaml
NameSpace\Contact:
    Contact:
        lastname: {Roe, Doe, Smith}
        firstname: {John, Jane, Paul}
        dob: {1980-01-01, 1981-01-01, 1982-01-01}
```
This will create 27 Contact objects.

I've seen that I could do something like:
```yaml
NameSpace\Contact:
    Contact_{Roe, Doe, Smith}:
        lastname: <current()>
        firstname: John
        dob: 1980-01-01
```
witch will reduce my inputs, but it's still a lot of work.

Is it possible to do such thing?

**Comment by theofidry** [comment]
>This will create 27 Contact objects.

So I guess you want to create all the possible fixtures with the input you gave is that right?

TL;DR: no.

You somehow have a way to do it, stores this matrix as [a parameter](https://github.com/nelmio/alice/blob/2.x/doc/fixtures-refactoring.md#parameters):

```yaml
parameters:
  contact_matrix:
    lastnames: [Roe, Doe, Smith}
    firstnames: [John, Jane, Paul]
    dob: [1980-01-01, 1981-01-01, 1982-01-01]
```

Then use a custom [Faker data provider](https://github.com/nelmio/alice/blob/2.x/doc/customizing-data-generation.md#custom-faker-data-providers) to be able to pick the right value and do something like this:

```yaml
NameSpace\Contact:
    Contact_{1..27}:
        lastname: <pickLastName(<{contact_matrix}>)>
        firstname: <pickFirstName(<{contact_matrix}>, $lastname)>
        dob: <pickDob(<{contact_matrix}>, $lastname, $firstname)>
```

So as you can see, you provider will have access to all the possible values, knows which row to look in and for firstname, as you know the possible values and the value of lastname, you can know which firstname pick.

It's the rough lines obviously and as you can see a bit of work. The reason being alice is not really meant to be used like that. That said, if you find something somewhat generic that can be re-used, we can consider merging it into the core or make a dedicate extension for alice.

**Comment by aledeg** [comment]
@theofidry Thank you. That makes almost perfect sense. I haven't though of using custom provider and I didn't know about the parameter feature.

I don't understand how ```<pickLastName(<{contact_matrix}>)>``` will work to return the appropriated value. I understand that I have to inject the parameters but how do I know which one I need to return?

**Comment by theofidry** [comment]
Actually giving it another though a simpler thing would be to make everything in the custom provider: you store the possible values in there and you have a method counter, so that the first call of `pickLastName()` and `pickFirstName()` will return `Roe` and `John`, the second call will return `Roe` and `Jane` and so on... it will not be re-usable, but it will work for this use case.

**Comment by aledeg** [comment]
I ended up doing what you suggested. It's not really important that it's not reusable for now since it's for a few related tests. If I need something more reusable, I'll rework it.
Thanks

**Comment by theofidry** [comment]
You're welcome :)
